pkg/controller/proxyconfig/validation.go

```diff
@@ -0,0 +1,346 @@
+		return "", fmt.Errorf("failed validating URI, no scheme for URI %q", uri)
+	}
+	host := parsed.Hostname()
+	if err := validateHost(host); err != nil {
+		return "", fmt.Errorf("failed validating URI %q: %v", uri, err)
+	}
+	if port := parsed.Port(); len(port) != 0 {
+		intPort, err := strconv.Atoi(port)
+		if err != nil {
+			return "", fmt.Errorf("failed converting port to integer for URI %q: %v", uri, err)
+		}
+		if err := validatePort(intPort); err != nil {
+			return "", fmt.Errorf("failed to validate port for URL %q: %v", uri, err)
+		}
+	}
+
+	return parsed.Scheme, nil
+}
+
+// validateHost validates if host is a valid IP address or subdomain in DNS (RFC 1123).
+func validateHost(host string) error {
+	errDomain := validateDomainName(host, false)
+	errIP := validation.IsValidIP(host)
+	if errDomain != nil && errIP != nil {
+		return fmt.Errorf("invalid host: %s", host)
+	}
+
+	return nil
+}
+
+// validatePort validates if port is a valid port number between 1-65535.
+func validatePort(port int) error {
+	invalidPorts := validation.IsValidPortNum(port)
+	if invalidPorts != nil {
+		return fmt.Errorf("invalid port number: %d", port)
+	}
+
+	return nil
+}
+
+// validateHTTPReadinessEndpoint validates an http readinessEndpoint endpoint.
+func validateHTTPReadinessEndpoint(endpoint string) error {
+	if err := validateHTTPReadinessEndpointWithRetries(endpoint, proxyProbeMaxRetries); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// validateHTTPReadinessEndpointWithRetries tries to validate an http
+// endpoint in a finite loop and returns the last result if it never succeeds.
+func validateHTTPReadinessEndpointWithRetries(endpoint string, retries int) error {
+	var err error
+	for i := 0; i < retries; i++ {
+		err = runHTTPReadinessProbe(endpoint)
+		if err == nil {
+			return nil
+		}
+	}
+
+	return err
+}
+
+// runHTTPReadinessProbe issues an http GET request to endpoint and returns
+// an error if a 2XX or 3XX http status code is not returned.
+func runHTTPReadinessProbe(endpoint string) error {
+	resp, err := http.Get(endpoint)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode >= http.StatusOK && resp.StatusCode < http.StatusBadRequest {
+		return nil
+	}
+
+	return fmt.Errorf("HTTP probe failed with statuscode: %d", resp.StatusCode)
+}
+
+// validateHTTPSReadinessEndpoint validates endpoint using certBundle as trusted CAs.
+func validateHTTPSReadinessEndpoint(certBundle []*x509.Certificate, endpoint string) error {
+	if err := validateHTTPSReadinessEndpointWithRetries(certBundle, endpoint, proxyProbeMaxRetries); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// validateHTTPSReadinessEndpointWithRetries tries to validate endpoint
+// by using certBundle as trusted CAs to create a TLS connection using a
+// finite loop based on retries, returning an error if it never succeeds.
+func validateHTTPSReadinessEndpointWithRetries(certBundle []*x509.Certificate, endpoint string, retries int) error {
```

Comment:
The old buggy behavior of `validateHTTPReadinessEndpointWithRetries()` is present here.

Reply:
ack